notebooks/proteochemometrics.py

```diff
@@ -0,0 +1,318 @@
+            arr = np.zeros((bitvect.GetNumBits(),), dtype=np.uint8)
+            ConvertToNumpyArray(bitvect, arr)
+
+            chembl_curie_to_idx[chembl_curie] = next(count)
+            tensors.append(torch.tensor(arr))
+            actual_chembl_curies.append(chembl_curie)
+
+    tensor = torch.stack(tensors)
+    if trainable:
+        representation = FeatureEnrichedEmbedding(tensor, shape=32)
+    else:
+        representation = Embedding.from_pretrained(tensor)
+    return RepresentationBackmap(representation, actual_chembl_curies, chembl_curie_to_idx)
+
+
+def get_protein_go_triples():
+    """Get triples with proteins as the subject and Gene Ontology terms as the objects."""
+    path = pystow.ensure("bio", "go", url=GO_URL)
+    df = pd.read_csv(
+        path, sep="\t", comment="!", usecols=[0, 1, 3, 4], names=["DB", "DB Object ID", "Relation", "GO ID"]
+    )
+    df = df[~df["Relation"].str.startswith("NOT")]
+    df = df[df["DB"] == "UniProtKB"]
+    df["subject"] = df["DB Object ID"].map(lambda s: f"uniprot:{s}")
+    df = df[["subject", "Relation", "GO ID"]].rename(columns={"Relation": "predicate", "GO ID": "object"})
+    return df
+
+
+def get_chemical_protein_triples():
+    """Get triples from ExCAPE-DB with chemicals as subjects and proteins as objects."""
+    excape_module = pystow.module("bio", "excape")
+    if (path := excape_module.join(name="excape_human_chembl_subset.tsv")).is_file():
+        df = pd.read_csv(path, sep="\t")
+        return df
+
+    df = excape_module.ensure_csv(
+        url=EXCAPE_URL,
+        read_csv_kwargs=dict(
+            sep="\t",
+            compression="xz",
+            usecols=["DB", "Original_Entry_ID", "Entrez_ID", "Activity_Flag", "Tax_ID"],
+            dtype=str,
+        ),
+    )
+    # keep only active relationships
+    df = df[df["Activity_Flag"] == "A"]
+    # slice down to human targets
+    df = df[df["Tax_ID"] == "9606"]
+    # keep only relationships qualified by ChEMBL, even though this way out of date in 2025
+    df = df[df["DB"] == "chembl20"]
+
+    df["object"] = df["Entrez_ID"].map(_uniprot_curie_from_entrez)
+
+    # throw away anything that can't be mapped to human uniprot
+    df = df[df["object"].notna()]
+
+    # turn into CURIEs
+    df["subject"] = df["Original_Entry_ID"].map(lambda s: f"chembl.compound:{s}")
+
+    df = df[["subject", "object"]].drop_duplicates()
+    df.to_csv(path, sep="\t", index=False)
+    return df
+
+
+def _uniprot_curie_from_entrez(s) -> str | None:
+    if pd.isna(s):
+        return None
+    uniprot_id = uniprot_client.get_id_from_entrez(s)
+    return uniprot_id and f"uniprot:{uniprot_id}"
+
+
+def main() -> None:
+    """Demonstrate using chemical representations for a subset of entities."""
+    target_dim = 32
+
+    device = torch.device("cpu")
+
+    # set this to true to enrich the chemical and protein features
+    # with additional learnable embeddings. Note that this makes
+    # training take a _lot_ longer
```

Comment:
With "a lot longer" you are referring to the number of iterations, or the iteration/step time?